vent/controller/control_module.py

```diff
@@ -6,6 +6,32 @@
 
 from vent.common.message import SensorValues, ControlSetting, Alarm, AlarmSeverity, ControlSettingName
 
+class RingBuffer:
```

Comment:
strongly recommend using deque here. list pops are O(n) where deques are O(1). You set a `maxlen` for a deque so you don't need to pop, just append to add.
https://docs.python.org/2/library/collections.html#collections.deque

```python
from collections import deque
dq = deque(maxlen=size)
dq.append(0)

```

Reply:
deque is threadsafe, no need to lock.

Reply:
never used this before, but sure, I'll recode this

Reply:
can you check? More specifically, I am not exactly sure about the relation of deque and lock.

Reply:
But the point of using a ring buffer is so that we can't get an infinitely growing list of logged values that eventually consumes all memory and crashes the system. Is deque a ring buffer?

Reply:
deque has a max length -- this is the case it was designed for, fast pushes and pops to a finite-length sequence. will look tonight

Reply:
Our use case was we wanted to keep the most recently logged items and overwrite the older ones if the buffer fills up. I haven't looked at ringbuffer implementation, but typically it would be an array with a two pointers, one to the first element and one to the last element. So push and pop would be O(1) if that was the implementation.

Reply:
yes yes that & some blocking memory management appears to be how deque is implemented: https://github.com/python/cpython/blob/a86b522d8f1c8b9c26b5550de221d2227158cf4d/Modules/_collectionsmodule.c#L90

Reply:
i think your use of the lock looks fine, deque should be thread-safe but locking while you replace it seems like a good idea.

Reply:
alright, then I'll merge this pull request, and clean up the code for the rest of the day